File: agent_engines/reserve.py
"""Reserve contracts — an asset (typically the battery or hospital backup)
committing to hold a slice of capacity for emergency use only, as opposed
to an active trade. Built to satisfy the contract's stated ownership
(Person 2 produces ReserveContract) but not yet exercised by anything else
in the pipeline — nobody currently calls this. If the demo scenario ends
up needing "hospital reserves its own battery, refuses to trade it away,"
this is the object that represents that commitment.
"""
import uuid
from datetime import datetime

from shared.contracts import ReserveContract


def create_reserve_contract(asset_id: str, reserved_kw: float, valid_until: datetime,
                             purpose: str) -> ReserveContract:
    return ReserveContract(
        contract_id=uuid.uuid4().hex[:8],
        asset_id=asset_id,
        reserved_kw=reserved_kw,
        valid_until=valid_until,
        purpose=purpose,
    )

File: agent_engines_rs/src/reserve.rs
//! Direct port of agent_engines/reserve.py. Not called by anything else
//! in the pipeline yet — same status as the Python version.

use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::contracts::ReserveContract;

pub fn create_reserve_contract(
    asset_id: String,
    reserved_kw: f64,
    valid_until: DateTime<Utc>,
    purpose: String,
) -> ReserveContract {
    ReserveContract {
        contract_id: Uuid::new_v4().simple().to_string()[..8].to_string(),
        asset_id,
        reserved_kw,
        valid_until,
        purpose,
    }
}
